# ADS 509 Module 1: APIs and Web Scraping

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## Important Note

This assignment requires you to have a version of Tweepy that is at least version 4. The latest version is 4.10 as I write this. Critically, this version of Tweepy is *not* on the upgrade path from Version 3, so you will not be able to simply upgrade the package if you are on Version 3. Instead you will need to explicitly install version 4, which you can do with a command like this: `pip install "tweepy>=4"`. You will also be using Version 2 of the Twitter API for this assignment. 

Run the below cell. If your version of Tweepy begins with a "4", then you should be good to go. If it begins with a "3" then run the following command, found [here](https://stackoverflow.com/questions/5226311/installing-specific-package-version-with-pip), at the command line or in a cell: `pip install -Iv tweepy==4.9`. (You may want to update that version number if Tweepy has moved on past 4.9. 

In [9]:
pip show tweepy

Name: tweepy
Version: 4.10.1
Summary: Twitter library for Python
Home-page: https://www.tweepy.org/
Author: Joshua Roesslein
Author-email: tweepy@googlegroups.com
License: MIT
Location: /Users/bikramgill/opt/miniconda3/envs/ADS502/lib/python3.9/site-packages
Requires: requests-oauthlib, requests, oauthlib
Required-by: 
Note: you may need to restart the kernel to use updated packages.


# Twitter API Pull

In [37]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup
from collections import defaultdict, Counter
import shutil

In [11]:
# Use this cell for any import statements you add
from os.path import exists
import random
import json


We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [12]:
from api_keys import api_key, api_key_secret, bearer_token

In [13]:
client = tweepy.Client(bearer_token,wait_on_rate_limit=True)

# Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull some of the followers @37chandler.
1. Explore response data, which gives us information about Twitter users. 
1. Pull the last few tweets by @37chandler.


In [14]:
handle = "37chandler"
user_obj = client.get_user(username=handle)

followers = client.get_users_followers(
    # Learn about user fields here: 
    # https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user
    user_obj.data.id, user_fields=["created_at","description","location",
                                   "public_metrics"]
)



Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users. 

In [15]:
num_to_print = 20

for idx, user in enumerate(followers.data) :
    following_count = user.public_metrics['following_count']
    followers_count = user.public_metrics['followers_count']
    
    print(f"{user.name} lists '{user.location}' as their location.")
    print(f" Following: {following_count}, Followers: {followers_count}.")
    print()
    
    if idx >= (num_to_print - 1) :
        break
    

Dave Renn lists 'None' as their location.
 Following: 42, Followers: 11.

Lionel lists 'None' as their location.
 Following: 202, Followers: 204.

Megan Randall lists 'None' as their location.
 Following: 141, Followers: 100.

Jacob Salzman lists 'None' as their location.
 Following: 562, Followers: 134.

twiter not fun lists 'None' as their location.
 Following: 221, Followers: 21.

Hariettwilsonincarnate lists 'None' as their location.
 Following: 218, Followers: 60.

Christian Tinsley lists 'None' as their location.
 Following: 2, Followers: 0.

Steve lists 'I'm over here.' as their location.
 Following: 1589, Followers: 33.

John O'Connor 🇺🇦 lists 'None' as their location.
 Following: 8, Followers: 1.

CodeGrade lists 'Amsterdam' as their location.
 Following: 2819, Followers: 421.

Cleverhood lists 'Providence, RI' as their location.
 Following: 2793, Followers: 3555.

Regina 🚶‍♀️🚲🌳 lists 'Minneapolis' as their location.
 Following: 2796, Followers: 3324.

Eric Hallstrom lists 'Mi

Let's find the person who follows this handle who has the most followers. 

In [16]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

WedgeLIVE
{'followers_count': 14157, 'following_count': 2222, 'tweet_count': 56072, 'listed_count': 218}


Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [17]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [18]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for username we have 37chandler
for id we have 33029025
for created_at we have 2009-04-18 22:08:22+00:00
for name we have John Chandler
for verified we have False
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for public_metrics we have {'followers_count': 193, 'following_count': 589, 'tweet_count': 994, 'listed_count': 3}
for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for location we have MN


#### Q: How many fields are being returned in the `response` object? 

#### A: 9 (code below)

In [19]:
count = 0

for field in response.data.items():
    count += 1
    
count

9

####  Q: Are any of the fields within the user object non-scalar? (I.e., more complicated than a simple data type like integer, float, string, boolean, etc.) 

####  A: From code below, created_at and public_metrics more complex data types

In [20]:
for field, value in response.data.items():
    print("{} type ---> {}".format(field, type(value)))

username type ---> <class 'str'>
id type ---> <class 'int'>
created_at type ---> <class 'datetime.datetime'>
name type ---> <class 'str'>
verified type ---> <class 'bool'>
profile_image_url type ---> <class 'str'>
public_metrics type ---> <class 'dict'>
description type ---> <class 'str'>
location type ---> <class 'str'>


#### Q: How many friends, followers, and tweets does this user have? 

#### A: From earlier code, followers - 193, following = 589 and tweet count = 994.

Although you won't need it for this assignment, individual tweets can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the fields that are available about Tweets.

In [21]:

response = client.get_users_tweets(user_obj.data.id)

# By default, only the ID and text fields of each Tweet will be returned
for idx, tweet in enumerate(response.data) :
    print(tweet.id)
    print(tweet.text)
    print()
    
    if idx > 10 :
        break

1568242374085869570
RT @depthsofwiki: lots of memes about speedy wikipedia editors — quick thread about what went down on wikipedia in the minutes after her de…

1568074978754703361
@DrLaurenWilson @leighradwood @MaritsaGeorgiou @Walgreens I could not possibly agree more with this sentiment. Compared to almost any other primary care I've received, they are great.

1567530169686196224
@DrLaurenWilson @MaritsaGeorgiou @Walgreens For those who have access to Curry Health Center on campus, you can get a bivalent booster in 15 minutes from their delightful staff.

1567511181526708224
RT @shes_the_maNN1: I can’t describe how ancient this makes me feel. https://t.co/a1IvELjOFY

1567510612665864193
RT @AngryBlackLady: this is hilarious

1566031636457725953
RT @MarkJacob16: With all the arguments over whether MAGA Republicans are fascists, I reread William Shirer’s “The Rise and Fall of the Thi…

1563737816219000832
RT @wonderofscience: The Milky Way galaxy and a phenomenon known as "airglow" s

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We've seen above how to pull a set of followers using `client.get_users_followers`. This function has a parameter, `max_results`, that we can use to change the number of followers that we pull. Unfortunately, we can only pull 1000 followers at a time, which means we will need to handle the _pagination_ of our results. 

The return object has the `.data` field, where the results will be found. It also has `.meta`, which we use to select the next "page" in the results using the `next_token` result. I will illustrate the ideas using our user from above. 


### Rate Limiting

Twitter limits the rates at which we can pull data, as detailed in [this guide](https://developer.twitter.com/en/docs/twitter-api/rate-limits). We can make 15 user requests per 15 minutes, meaning that we can pull $4 \cdot 15 \cdot 1000 = 60000$ users per hour. I illustrate the handling of rate limiting below, though whether or not you hit that part of the code depends on your value of `handle`.  


In the below example, I'll pull all the followers, 25 at a time. (We're using 25 to illustrate the idea; when you do this set the value to 1000.) 

In [22]:
handle_followers = []
pulls = 0
max_pulls = 100
next_token = None

while True :

    followers = client.get_users_followers(
        user_obj.data.id, 
        max_results=25, # when you do this for real, set this to 1000!
        pagination_token = next_token,
        user_fields=["created_at","description","location",
                     "entities","name","pinned_tweet_id","profile_image_url",
                     "verified","public_metrics"]
    )
    pulls += 1
    
    for follower in followers.data : 
        follower_row = (follower.id,follower.name,follower.created_at,follower.description)
        handle_followers.append(follower_row)
    
    if 'next_token' in followers.meta and pulls < max_pulls :
        next_token = followers.meta['next_token']
    else : 
        break
        



## Pulling Twitter Data for Your Artists

Now let's take a look at your artists and see how long it is going to take to pull all their followers. 

In [23]:
artists = dict()

for handle in ['robynkonichiwa','cher'] : 
    user_obj = client.get_user(username=handle,user_fields=["public_metrics"])
    artists[handle] = (user_obj.data.id, 
                       handle,
                       user_obj.data.public_metrics['followers_count'])
    

for artist, data in artists.items() : 
    print(f"It would take {data[2]/(1000*15*4):.2f} hours to pull all {data[2]} followers for {artist}. ")
    

It would take 5.94 hours to pull all 356373 followers for robynkonichiwa. 
It would take 67.41 hours to pull all 4044752 followers for cher. 


Depending on what you see in the display above, you may want to limit how many followers you pull. It'd be great to get at least 200,000 per artist. 

As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [24]:
# Make the "twitter" folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then "unlink" it. Then create a new one.

if not os.path.isdir("twitter") : 
    #shutil.rmtree("twitter/")
    os.mkdir("twitter")

In this following cells, build on the above code to pull some of the followers and their data for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In addition to creating a file that only has follower IDs in it, you will create a file that includes user data. From the response object please extract and store the following fields: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 

One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. I've included some example code to do that below the stub. 

In [25]:
num_followers_to_pull = 200*1000 # feel free to use this to limit the number of followers you pull.

In [26]:
# Modify the below code stub to pull the follower IDs and write them to a file. 

handles = ['BrunoMars','rihanna']

whitespace_pattern = re.compile(r"\s+")

user_data = dict() 
followers_data = dict()

for handle in handles :
    user_data[handle] = [] # will be a list of lists
    followers_data[handle] = [] # will be a simple list of IDs


# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()

for handle in handles :
    
    print("Working on handle: {}".format(handle))
    
    
    ## Get User ID
    user_obj = client.get_user(username=handle)
    handle_id = user_obj.data.id
    
    # Create the output file names 
    # Files will be cleared before running API in next cell block
    
    followers_output_file = handle + "_followers.txt"
    user_data_output_file = handle + "_follower_data.txt"
    

    
    ## Clear files if code has already been run earlier. 
    ##Will allow paginator to append user id's later without duplciating data. 
    files = [followers_output_file, user_data_output_file]
    for file in files:
        path = os.path.join("twitter", file)
        
        if exists(path):
            with open(path, 'r+') as doc:
                doc.truncate(0)     
    
    
    ## Query follower info for handle via pagination
    
    for response in tweepy.Paginator(client.get_users_followers, handle_id, 
                                     user_fields=["username","name","id", "location","description","public_metrics"],
                                     max_results=1000, limit=110):
        

        for followers in response.data:
            
            ## Write user names of followers to dict and file
            followers_data[handle].append(followers.username)
            
            with open(os.path.join("twitter", followers_output_file), 'a') as file:
                file.write("{}\n".format(followers.username))
                
                
            
            ## Write user info of followers to dict and file
            follower_row = [str(followers.username),
                           str(followers.name),
                           str(followers.id),
                           str(followers.location),
                           str(followers.public_metrics['followers_count']),
                           str(followers.public_metrics['following_count']),
                           str(re.sub(r"\s+"," ",followers.description).strip())]
            
            user_data[handle].append(follower_row)
            
            
            with open(os.path.join("twitter", user_data_output_file), 'a') as file:
                # Create a tab delimited version of each follower row
                file.write('\t'.join(follower_row) + '\n')
                                     
                            


        
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print(end_time - start_time)

Working on handle: BrunoMars


Rate limit exceeded. Sleeping for 785 seconds.
Rate limit exceeded. Sleeping for 765 seconds.
Rate limit exceeded. Sleeping for 750 seconds.
Rate limit exceeded. Sleeping for 763 seconds.
Rate limit exceeded. Sleeping for 756 seconds.
Rate limit exceeded. Sleeping for 764 seconds.
Rate limit exceeded. Sleeping for 764 seconds.
Rate limit exceeded. Sleeping for 764 seconds.


Working on handle: rihanna


Rate limit exceeded. Sleeping for 763 seconds.
Rate limit exceeded. Sleeping for 764 seconds.
Rate limit exceeded. Sleeping for 764 seconds.
Rate limit exceeded. Sleeping for 764 seconds.
Rate limit exceeded. Sleeping for 761 seconds.
Rate limit exceeded. Sleeping for 762 seconds.
Rate limit exceeded. Sleeping for 764 seconds.


3:44:59.695910


In [32]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.

"""
# This won't work in a tab-delimited text file.

clean_description = re.sub(r"\s+"," ",tricky_description).strip()
clean_description

'Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well.'

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [28]:
artists = {'brunomars':"https://www.azlyrics.com/b/brunomars.html",
           'rihanna':"https://www.azlyrics.com/r/rihanna.html"} 
# we'll use this dictionary to hold both the artist name and the link on AZlyrics

## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

#### Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

#### A: Based on the robotx.txt file, it looks like only lyricsdb and song path access is disallowed, whilst all others are allowed. 


In [29]:
# Let's set up a dictionary of lists to hold our links
lyrics_pages = defaultdict(list)

for artist, artist_page in artists.items() :
    # request the page and sleep
    r = requests.get(artist_page)
    time.sleep(5 + 10*random.random())

    # now extract the links to lyrics pages from this page
    # store the links `lyrics_pages` where the key is the artist and the
    # value is a list of links. 
    
    urls = []
    soup = BeautifulSoup(r.text, 'html.parser')
    
    for link in soup.find_all('a'):
        if ("lyrics/" + artist) in str(link.get('href')):
            urls.append('https://www.azlyrics.com/' + str(link.get('href')))

    lyrics_pages[artist] = urls

Let's make sure we have enough lyrics pages to scrape. 

In [32]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [33]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

For brunomars we have 79.
The full pull will take for this artist will take 0.22 hours.
For rihanna we have 160.
The full pull will take for this artist will take 0.44 hours.


## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [64]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    link = str(link)
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = link.replace(".html","")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    # additional processing remove unwanted info
    name = name.replace("httpswww_azlyrics_com_", "")
    
    return(name)


In [38]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.

if os.path.isdir("lyrics") : 
    shutil.rmtree("lyrics/")

os.mkdir("lyrics")

In [ ]:
url_stub = "https://www.azlyrics.com" 
start = time.time()

total_pages = 0 

for artist in lyrics_pages :
   
    # Use this space to carry out the following steps: 
    
    # 1. Build a subfolder for the artist
    
    if os.path.isdir("lyrics/" + artist):
        shutil.rmtree("lyrics/"+artist)
        
    os.mkdir("lyrics/"+artist)
    
    # 2. Iterate over the lyrics pages   
    # 3. Request the lyrics page. 
        # Don't forget to add a line like `time.sleep(5 + 10*random.random())`
        # to sleep after making the request
    # 4. Extract the title and lyrics from the page.
    # 5. Write out the title, two returns ('\n'), and the lyrics. Use `generate_filename_from_url`
    #    to generate the filename. 
    
    ## Combining all steps above:
    
    for lyrics in lyrics_pages[artist]:
        r = requests.get(lyrics)
        time.sleep(5 + 10*random.random())
        
        soup = BeautifulSoup(r.content, "html.parser")
        
        ## Get Song Title
        title = soup.select_one("h1").get_text(strip=True, separator="\n").replace(" lyrics", "").strip('\"')
        
        ## Get Lyrics Text
        lyric = soup.select_one(".ringtone ~ div").get_text(strip=True, separator="\n")
        
        
        ## Generate file name and save to file 
        filename = generate_filename_from_link(lyrics)
        
        with open(os.path.join("lyrics", artist, filename), 'w') as file:
            file.write(title + '\n\n' + lyric)
        
        print("Created and closed --> {}".format(filename))

Created and closed --> brunomars_somewhereinbrooklyn.txt
Created and closed --> brunomars_theotherside.txt
Created and closed --> brunomars_countonme.txt
Created and closed --> brunomars_talkingtothemoon.txt
Created and closed --> brunomars_grenade.txt
Created and closed --> brunomars_justthewayyouare.txt
Created and closed --> brunomars_ourfirsttime.txt
Created and closed --> brunomars_runawaybaby.txt
Created and closed --> brunomars_thelazysong.txt
Created and closed --> brunomars_marryyou.txt
Created and closed --> brunomars_talkingtothemoon.txt
Created and closed --> brunomars_liquorstoreblues.txt
Created and closed --> brunomars_countonme.txt
Created and closed --> brunomars_theotherside.txt
Created and closed --> brunomars_younggirls.txt
Created and closed --> brunomars_lockedoutofheaven.txt
Created and closed --> brunomars_gorilla.txt
Created and closed --> brunomars_treasure.txt
Created and closed --> brunomars_moonshine.txt
Created and closed --> brunomars_wheniwasyourman.txt


Created and closed --> rihanna_numb.txt
Created and closed --> rihanna_pouritup.txt
Created and closed --> rihanna_loveeeeeeesong.txt
Created and closed --> rihanna_jump.txt
Created and closed --> rihanna_rightnow.txt
Created and closed --> rihanna_whatnow.txt
Created and closed --> rihanna_stay.txt
Created and closed --> rihanna_nobodysbusiness.txt
Created and closed --> rihanna_lovewithouttragedymothermary.txt
Created and closed --> rihanna_getitoverwith.txt
Created and closed --> rihanna_noloveallowed.txt
Created and closed --> rihanna_lost-in-paradise.txt
Created and closed --> rihanna_halfofme.txt
Created and closed --> rihanna_diamondsdaveaud100extended.txt
Created and closed --> rihanna_diamondsgregorsaltodowntemporemix.txt
Created and closed --> rihanna_towardsthesun.txt
Created and closed --> rihanna_asrealasyouandme.txt
Created and closed --> rihanna_dancinginthedark.txt
Created and closed --> rihanna_consideration.txt
Created and closed --> rihanna_jamesjoint.txt
Created and

In [ ]:
print(f"Total run time was {round((time.time() - start)/3600,2)} hours.")

---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [ ]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [ ]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

In [ ]:
for artist in artist_handles :
    follower_file = artist + "_followers.txt"
    follower_data_file = artist + "_followers_data.txt"
    
    ids = open("twitter/" + follower_file,'r').readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + follower_data_file,'r') as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({follower_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[3])            
                description_words.extend(words(line[6]))
            except :
                pass
    
        

        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        
        print("")
        print("-"*40)
        print("")
    

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [ ]:
artist_folders = os.listdir("lyrics/")
artist_folders = [f for f in artist_folders if os.path.isdir("lyrics/" + f)]

for artist in artist_folders : 
    artist_files = os.listdir("lyrics/" + artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open("lyrics/" + artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")
